### PR TITLE
Enabled Model.getOps() method

### PR DESCRIPTION
**Pull request by spran180**
### Details:
 - *Implemented Node.js binding for OpenVino's `ModelWrap` and `NodeWrap` classes. *
 - *Added methods to expose model and node functionalities to JavaScript. *

### Tickets:
 - *#25405*

### CODE REVIEW

File: src/bindings/js/node/tests/model.test.js [hunk 1]
```diff
@@ -111,4 +111,33 @@ describe('Model.getOutputSize()', () => {
   it('should return 1 for the default model', () => {
     assert.strictEqual(model.getOutputSize(), 1, 'Expected getOutputSize to return 1 for the default model');
   });
+});
+
+describe('Node.js Model.getOps()', () => {
+  it('should return array of Node operations', () => {
+    const model = new addon.ModelWrap();
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/tests/model.test.js:118]: You cannot get `Model` instance using constructor. `ModelWrap` doesn't have constructor. You can get it by call `Core.readModel` method.

File: src/bindings/js/node/tests/model.test.js [hunk 2]
```diff
@@ -111,4 +111,33 @@ describe('Model.getOutputSize()', () => {
...
+      "Subtract",
+      "Transpose",
+    ];
+
+    assert.strictEqual(modelOperators.length, 14, 'Expected 14 operations in the model');
+    expectedOps.forEach(op => {
+      assert(modelOperators.includes(op), `Expected operation ${op} to be in the model`);
+    });
+  });
 });
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/tests/model.test.js:143]: Empty line at the end of the file is required

**Review by vishniakov-nikolai — CHANGES_REQUESTED** [review]
Pranav, thank you for your contribution! 
Please pay attention to the remarks and ask question if something isn't clear.

File: src/bindings/js/node/include/node_wrap.hpp [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        * @brief Constructs NodeWrap from the Napi::CallbackInfo.
+        * @param info contains passed arguments. Can be empty.
+        */
+        static Napi::Object Init(Napi::Env env, Napi::Object exports);
+        NodeWrap(const Napi::CallbackInfo& info);
+        
+    
+    private:
+        static Napi::FunctionReference constructor;
+        ov::Node* node;
```
[the remark this answers is not in the document]
Comment (spran180) [reply src/bindings/js/node/include/node_wrap.hpp:17]: Since the ov::node is managed by the OpenVINO library, there is no need for us to handle its memory. However, if we wish to manage the object's lifetime then it would be store directly (without pointer).

**Comment by spran180** [comment]
Is it necessary to set a name for ov::Node during initialization?

**Comment by vishniakov-nikolai** [comment]
> Is it necessary to set a name for ov::Node during initialization?

There is no need to create ov::Node from js bindings.
You get nodes array by call `getOps` method. You already have original node and can bind name request to it.
From js side you cannot change name value.

**Comment by spran180** [comment]
Hey @vishniakov-nikolai,

I've addressed the issues in code pls check it and let me know if further changes are required.

File: src/bindings/js/node/include/node_wrap.hpp [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+     */
+    NodeWrap(const Napi::CallbackInfo& info);
+
+    static Napi::Function get_class(Napi::Env env);
+
+private:
+    static Napi::FunctionReference constructor;
+    std::shared_ptr<ov::Node> node;
+
+    Napi::Value get_name(const Napi::CallbackInfo& info)
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/include/node_wrap.hpp:20]: You can move it to `public` part

File: src/bindings/js/node/lib/addon.ts [hunk 5]
```diff
@@ -284,8 +284,18 @@ interface Model {
    * It gets all the model outputs as an array
    */
   outputs: Output[];
+  /**
+   * It returns the array of NodeWrap objects.
+   */
+  get_ops(): NodeWrap[];
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/lib/addon.ts:290]: At JS side please use `camelCase` for the methods and variables names (but still `snake_case` for c++)

File: src/bindings/js/node/src/model_wrap.cpp [hunk 6]
```diff
@@ -25,6 +26,7 @@ Napi::Function ModelWrap::get_class(Napi::Env env) {
                         InstanceMethod("setFriendlyName", &ModelWrap::set_friendly_name),
                         InstanceMethod("getFriendlyName", &ModelWrap::get_friendly_name),
                         InstanceMethod("getOutputShape", &ModelWrap::get_output_shape),
+                        InstanceMethod("getOps", &ModelWrap::get_ops),
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/src/model_wrap.cpp:29]: Here is correct

File: src/bindings/js/node/include/model_wrap.hpp [hunk 7]
```diff
@@ -109,6 +109,13 @@ class ModelWrap : public Napi::ObjectWrap<ModelWrap> {
      */
     Napi::Value get_output_shape(const Napi::CallbackInfo& info);
 
+    /**
+     * @brief Function to access operations of the model
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/include/model_wrap.hpp:113]: Just use info from python api docs https://docs.openvino.ai/2024/api/ie_python_api/_autosummary/openvino.Model.html#openvino.Model.get_ops
```suggestion
     * @brief Return ops used in the model.
```

File: src/bindings/js/node/src/node_wrap.cpp [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+Napi::FunctionReference NodeWrap::constructor;
+
+Napi::Function NodeWrap::get_class(Napi::Env env) {
+    return DefineClass(env, "NodeWrap", {InstanceMethod("getName", &NodeWrap::get_name)});
+}
+
+NodeWrap::NodeWrap(const Napi::CallbackInfo& info) : Napi::ObjectWrap<NodeWrap>(info) {
+    Napi::Env env = info.Env();
+    std::vector<std::string> allowed_signatures = {"NodeWrap(Napi::External<ov::Node>)", "NodeWrap()"};
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/src/node_wrap.cpp:17]: Allowed signatures it's array that fill fulfill by validate method. It needs to show allowed signatures in the case of miss much.  There is no need to add anything to it here.
```suggestion
    std::vector<std::string> allowed_signatures;
```

File: src/bindings/js/node/src/node_wrap.cpp [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+    try {
+        if (ov::js::validate<Napi::External<ov::Node>>(info, allowed_signatures)) {
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/src/node_wrap.cpp:20]: To use `validate` method with `ov::Node` you have to extend https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/js/node/src/type_validation.cpp#L104 validation overloads.

File: src/bindings/js/node/tests/model.test.js [hunk 10]
```diff
@@ -111,4 +111,31 @@ describe('Model.getOutputSize()', () => {
...
+    result.forEach(op => {
+      assert.strictEqual(typeof op.get_name, 'function', 'each item should have get_name method');
+    });
+  });
+
+  it('should return the expected operation', () => {
+    const result = model.get_ops();
+
+    const modelOperators = result.map(op => op.get_name().split('_')[0]);
+    expected_ops = [
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/tests/model.test.js:131]: Variable definition should have `const` or `let`

File: src/bindings/js/node/tests/model.test.js [hunk 11]
```diff
@@ -111,4 +111,31 @@ describe('Model.getOutputSize()', () => {
...
+      "Subtract",
+      "Transpose",
+    ];
+
+    assert.strictEqual(modelOperators.length, 14, 'Expected 14 operations in the model');
```
Comment (vishniakov-nikolai) [inline src/bindings/js/node/tests/model.test.js:136]: 14 is 'magic number'. Specify the name for it.

**Review by vishniakov-nikolai — COMMENTED** [review]
Good job! But there are still remarks.

**Comment by mlukasze** [comment]
hey @spran180 thanks for update for PR
could you resolve conflict as well, please?

**Comment by spran180** [comment]
Hey @vishniakov-nikolai,

I have resolved the issues and conflicts in the files. Could you please review them?

File: src/bindings/js/node/src/model_wrap.cpp [hunk 12]
```diff
@@ -171,3 +173,16 @@ Napi::Value ModelWrap::get_output_shape(const Napi::CallbackInfo& info) {
         return info.Env().Undefined();
     }
 }
+
+Napi::Value ModelWrap::get_ops(const Napi::CallbackInfo& info) {
+    Napi::Env env = info.Env();
```
Comment (almilosz) [inline src/bindings/js/node/src/model_wrap.cpp:178]: Here please check if no arguments were passed using `ov::js::validate`

File: src/bindings/js/node/src/model_wrap.cpp [hunk 13]
```diff
@@ -171,3 +173,16 @@ Napi::Value ModelWrap::get_output_shape(const Napi::CallbackInfo& info) {
...
+    std::vector<std::shared_ptr<ov::Node>> ops = _model->get_ops();
+
+    Napi::Array result = Napi::Array::New(env, ops.size());
+    for (size_t i = 0; i < ops.size(); ++i) {
```
Comment (almilosz) [inline src/bindings/js/node/src/model_wrap.cpp:182]:
```suggestion
    for (uint32_t i = 0; i < ops.size(); ++i) {
```
`size_t` may cause errors/warnings on some systems

File: src/bindings/js/node/src/model_wrap.cpp [hunk 14]
```diff
@@ -171,3 +173,16 @@ Napi::Value ModelWrap::get_output_shape(const Napi::CallbackInfo& info) {
...
+        Napi::Object nodeWrap = NodeWrap::get_class(env).New({Napi::External<std::shared_ptr<ov::Node>>::New(env, &ops[i])});
```
Comment (almilosz) [inline src/bindings/js/node/src/model_wrap.cpp:183]: If you create js object like this, you will kind of create a new identical class and you won't be able to compare them using `instanceof` operator. 
To avoid that you can use previously created `Napi::Function` from `get_class`. It may look similar to this:
https://github.com/openvinotoolkit/openvino/blob/dc3eaf0a2b816fc32a59e79455bce33ec54f535c/src/bindings/js/node/src/helper.cpp#L255-L264

**Review by almilosz — COMMENTED** [review]
Hi spran180,
I really appreciate the work you've already put into this PR. Here a few more comments ;)

File: src/bindings/js/node/src/model_wrap.cpp (hunk 12 above)
Comment (spran180) [reply src/bindings/js/node/src/model_wrap.cpp:178]: hey @almilosz, 

Is it necessary to check using `ov::js::validate`?
I thing we can also check validation directly in function like this-
```suggestion
    Napi::Env env = info.Env();
    
    if (info.Length() != 0) {
        reportError(env, "Invalid argument. Expected no arguments.");
        return env.Undefined();
    }
```

Comment (spran180) [reply src/bindings/js/node/src/model_wrap.cpp:178]: hey @almilosz, 

Is it necessary to check using `ov::js::validate`?
I thing we can also check validation directly in function like this-
```suggestion
    Napi::Env env = info.Env();
    
    if (info.Length() != 0) {
        reportError(env, "Invalid argument. Expected no arguments.");
        return env.Undefined();
    }
```

Comment (almilosz) [reply src/bindings/js/node/src/model_wrap.cpp:178]: It will work but one of the purpose of using `ov::js::validate` is to ensure that error messages are consistent across different locations

**Comment by spran180** [comment]
Hey @almilosz, 

I have addressed the changes pls review it once.

File: src/bindings/js/node/src/addon.cpp [hunk 15]
```diff
@@ -40,6 +41,9 @@ Napi::Object init_module(Napi::Env env, Napi::Object exports) {
     init_class(env, exports, "Output", &Output<ov::Node>::get_class, addon_data->output);
     init_class(env, exports, "ConstOutput", &Output<const ov::Node>::get_class, addon_data->const_output);
     init_class(env, exports, "PartialShape", &PartialShapeWrap::get_class, addon_data->partial_shape);
+    
+    Napi::Function node_prototype = NodeWrap::get_class(env);
+    addon_data->node = Napi::Persistent(node_prototype);
```
Comment (almilosz) [inline src/bindings/js/node/src/addon.cpp:46]:
```suggestion
    addon_data->node = Napi::Persistent(NodeWrap::get_class(env));
```

File: src/bindings/js/node/src/node_wrap.cpp [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+Napi::Value NodeWrap::get_name(const Napi::CallbackInfo& info) {
+    Napi::Env env = info.Env();
+
+    if (!ov::js::validate<>(info)) {
+        reportError(env, "Invalid argument. Expected no arguments.");
+        return env.Undefined();
+    }
+
+    return Napi::String::New(env, node->get_name());
```
Comment (almilosz) [inline src/bindings/js/node/src/node_wrap.cpp:42]:
```suggestion
    return Napi::String::New(env, _node->get_name());
```

File: src/bindings/js/node/src/addon.cpp [hunk 17]
```diff
@@ -15,6 +15,7 @@
 #include "node/include/preprocess/preprocess.hpp"
 #include "node/include/tensor.hpp"
 #include "openvino/openvino.hpp"
+#include "node/include/node_wrap.hpp"
```
Comment (almilosz) [inline src/bindings/js/node/src/addon.cpp:18]: Please install `clang` and run `make -j44 clang_format_fix_all`.
This line will be probably moved up

File: src/bindings/js/node/include/type_validation.hpp [hunk 18]
```diff
@@ -89,6 +96,12 @@ struct InputParameters {
 
 template <typename... Ts>
 bool validate(const Napi::CallbackInfo& info) {
+
+    if constexpr (sizeof...(Ts) == 0) {
```
Comment (almilosz) [inline src/bindings/js/node/include/type_validation.hpp:100]: Is this change necessary?

**Review by almilosz — CHANGES_REQUESTED** [review]
Please make sure that the code compiles and run tests from _openvino/src/bindings/js/node/_ using `npm run test`
In case of any questions/troubles you can reach me here or on Discord `_almilosz`

**Comment by almilosz** [comment]
Hi, @spran180 
Are you still working on this? :)

**Comment by vishniakov-nikolai** [comment]
Hi @spran180!
I ping you, because this PR is going to be closed. Would you share it for someone else?
